doc/adr004-effect-system.md

```diff
@@ -38,150 +36,292 @@ simplicity.
+An effect can also have three different forms: concrete, arrow and variable.
 
+```typescript
+/*
+ * The effect of a Quint expression, regarding which state entities are read
+ * and/or updated
+ */
+export type Effect =
+  | ConcreteEffect
+  | ArrowEffect
+  | EffectVariable
 ```
-Effects E ::= Read[vars] & Update[vars] | (E0, ..., EN) => E
+
+The most basic representation of effects is a concrete effect, and it is
+structured as a list of effect components. An effect component has a kind and an
+entity. The kind can be `read`, `update` or `temporal`, and the entity is the
+entity the effect acts upon.
+
+```typescript
+/* Kinds for concrete effect components */
+export type ComponentKind = 'read' | 'update' | 'temporal'
+
+/* A concrete effect component, specifying a kind and the entity it acts upon */
+export interface EffectComponent { kind: ComponentKind, entity: Entity }
+
+/* Concrete atomic effects specifying which kinds of effects act upon which entities */
+export interface ConcreteEffect { kind: 'concrete', components: EffectComponent[] }
+
 ```
 
-Other forms are actually sugaring for these two. Effects like `Update['x'] &
-Update['y']` are in an invalid format and should be simplified before applying
-any other transformation. See equivalence rules below.
+The effect `Read['x', 'y'] & Update[v]` would be represented as:
 
-The motivation for this form is to help writing effect signatures for operators
-that care only about the read or the update part of some effect. For example,
-the `or` operator representing disjunction in Quint takes expressions with the
-identical updates, but doesn't have any restrictions on the read part. Ensuring
-this normal form allows us to write its signature as `(Read[r1] & Update[u],
-Read[r2] & Update[u]) => Read[r1 ∪ r2] & Update[u]`.
+```typescript
+{
+  kind: 'concrete',
+  components: [
+    { kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: 'x', reference: 1n }, { name: 'y', reference: 2n }] } },
+    { kind: 'update', entity: { kind: 'variable', name: 'v' } },
+  ],
+}
+```
 
-### Equivalence rules
+Effect signatures for operators may depend on its parameters. For that purpose,
+we also define arrow effects, which carry a list of effects for the respective
+parameters of an operator and the resulting effect.
 
-These are some equivalence rules to be used alongside unification, but that
-don't require any substitution. These are applied in a simplification process
-with the goal of reaching the normal form. Equivalence between `E1` and `E2` is
-expressed by `E1 ≡ E2`, and the equivalence symbol `≡` has the lowest precedence
-on this system.
+```typescript
+/* Arrow effects for expressions with effects depending on parameters */
+export interface ArrowEffect { kind: 'arrow', params: Effect[], result: Effect }
+```
 
+Just like with entities, an effect can also be a variable, which can be
+substituted with another effect during inferrence. Those are the most general
```

Comment:
I think the phasing 'an effect can also be a variable' sets up a point of possoble conceptual and terminological confusion. Rather, there is a class of variables that range over effects (and a different class that range over "entities", and "state variables" aren't actually variables that range over anything, lol). So how about something like:

```suggestion
These is a separate class of variables that range over effects, which can be
substituted with another effect during inference. Those are the most general
```